### DISCUSSION TITLE
Cannot install medspacy and scispacy

### DISCUSSION THREAD

**Discussion by standenman** [medspacy/medspacy#238] [General]
I have wasted most of my day on this and I am at the end of my rope.  I  have a python virtual environment on my windows machine.  I cannot get scispacy and medspacy to install on my life.  I get an error that the nmslib wheel will not install.  Trying to install nmslib does not help because it will not install.  I have tired every work around I can find on the web.  I am about at the conclusion that medspacy simply will not work on a windows machine.

**Comment by burgersmoke** [comment]
Hi @standenman I am sorry that you are frustrated, so let's see if we can help.  medspacy certainly does install and work on Windows machines.  We have continuous integration which runs tests and `pip installs` on MacOS, Ubuntu, and Windows which all pass.  You can peek at these if you think they might help to shed any light on what is happening for you.  Most of us who work on medspacy use Windows as the primary platform.

To help figure out this situation, I propose we first take `scispacy` out of the equation and perhaps you could tell us a bit more about the problems you are seeing:

1. When installing medspacy only, which Python version are you running?  I ask since we do not support older versions of Python which are deprecated or some of the very latest Python interpreter versions (due to dependencies)
2. What commands are you running when you are trying just to run medspacy?
3. Can you share the output you see from this command?

**Reply by standenman** [reply]
Thanks for your reply.  I am able to get medspacy imported, but the code has two problems:   there are only three pipelines being loaded.  And I am not getting any "output" - just the text with no markings.  Here is my code:

"import medspacy
import sys
import spacy
import en_core_web_sm
from medspacy.visualization import visualize_ent
nlp = medspacy.load("en_core_web_sm")


# Load the desired medSpaCy model
nlp = medspacy.load()

#text = """
#Stanley Denman is a 60 y.o. male with a  history of fatigue for 10 years, placed on testosterone in 2020 and developed severe substernal chest pain, drove himself to Doctor's hospital and Dr. Mamood Ali placed LAD stent. Unforturnately, he had LVEF 25%, followed and had recurrent chest cramping, had another cardiac cath showing stent in the posterior artery. After his second stent he had ICD placed October 2021. He was told he had CHF at that time. He wants second opinion about his prognosis and his treatment plan. Currently he has severe DOE; can walk 1/2 block with his dog but has severe shortness of breath and heart palpitations- hard heart beating. He agrees to take his medications and has been compliant. After ICD he had chest pain and swelling. He had to be diuresed prior to receiving the ICD. He has gained 30# since MI and is not on testosterone. He is not on GDMT at this time. He is very upset and it is difficult for him to hear my discussion. Has severe muscle aching in hands and feet on lipitor. Continues to have pain on crestor. Albuterol HFA inhaler 90 mcg/actuation INHALE TWO PUFFS BY MOUTH EVERY 6 HOURS AS NEEDED FOR WHEEZING (Patient taking differently: prn.).  Drug 


print(nlp.pipe_names)

doc = nlp(text)
visualize_ent(doc)"
![Screenshot (72)](https://github.com/medspacy/medspacy/assets/17820082/5646e0b0-cc48-4484-8d7f-f883f6ed3017)

**Comment by burgersmoke** [comment]
@standenman Yes, this is expected behavior.  I got confused as the original discussion title here refers to installation issues, but I think that this comes down to expectations of what medspacy is and what it provides out-of-the-box.  

As @abchapman93  pointed out in this other discussion, the code you are using here is using the default spacy model (`en_core_web_sm`)  which will not extract clinical concepts such as medications, signs, symptoms, etc out of the box: 

https://github.com/medspacy/medspacy/discussions/235#discussioncomment-6019458

Rather, you have 3 options as I see them:
1. Use `scispacy` as suggested (since `en_core_web_sm` does not contain clinical concepts)
2. Create your own QuickUMLS database using a full UMLS distribution (note we do not make this available by default as we understand this would violate the user license of UMLS).  More info on how to do this is here under #3 of Installation: https://github.com/medspacy/quickumls 
3. Create your own rules

**Comment by burgersmoke** [comment]
Also, if you are encountering issues installing both scispacy and medspacy into an environment, please share more information about that.  We install both (and many other off the shelf language models) as part of our continuous integration and they are installing and running without error.  

In case it is useful, this is the version which our continuous integrations installs:
`pip install https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/releases/v0.3.0/en_core_sci_sm-0.3.0.tar.gz`

**Reply by standenman** [reply]
Thank you again for your prompt and detailed answer.  Sorry I was a little vague.  We regard to my problems installing scispacy.  "Pip install scispacy"  fails at "Encountered error while trying to install package. "encountered error trying to install package nmslib".  Installing nmslib seperately also fails.  Using pip install https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/releases/v0.3.0/en_core_sci_sm-0.3.0.tar.gz in my venv does not throw any errors, but a pip list won't find it, and when I try to run the code in VS Code I get the following: 

"PS C:\Users\stand\venv> & c:/Users/stand/venv/Scripts/python.exe "c:/Users/stand/venv/import Medspacy.py"
Traceback (most recent call last):
  File "c:\Users\stand\venv\import Medspacy.py", line 2, in <module>
    import scispacy
ModuleNotFoundError: No module named 'scispacy'"

**Reply by burgersmoke** [reply]
OK, so this is an issue with installing scispacy and not related to medspacy.  However, I would still like to help you, so let's get back to my original questions for installation:

Which version of Python are you using in your environment?

While I am not a scispacy developer, I just googles "scispacy nmslib" and found some relevant articles right away.  This one is dated 2020 but it seems that only versions of Python up to 3.8 are supported by scispacy.  Specifically someone using Python 3.9 was getting failures  This might be different now, but it looks very similar to what you are encountering:

https://github.com/allenai/scispacy/issues/291

**Reply by standenman** [reply]
Well I thought medspacy was built on top of scispacy - that I need to scispacy.  And I am only getting 3 pipelines from my medspacy install.   Thought that there were more.  But if I don't need scispacy I would rather spare the aggravation.  I am using python 3.11.3

**Reply by burgersmoke** [reply]
I think this is the problem.  I will leave it to you to do the research on which version(s) of Python are supported by scispacy, but from what I read, I believe 3.11.x is too new.  I would probably recommend Python 3.8.x to ensure compatibility.

Since this is a scispacy issue (not medspacy) and I think this will resolve the problem, I am inclined to close this issue.

**Reply by burgersmoke** [reply]
I did a quick search.  It appears this is a known issue in scispacy that Python 3.11.x is not supported.  I will close this issue:

https://github.com/allenai/scispacy/issues/473
